stan/math/rev/functor/cvodes_integrator_adjoint.hpp

```diff
@@ -0,0 +1,829 @@
+
+    Eigen::Map<Eigen::VectorXd>(dy_dt, N_) = dy_dt_vec;
+  }
+
+  /*
+   * Calculate the adjoint sensitivity RHS for varying initial conditions
+   * and parameters
+   *
+   * Equation 2.23 in the cvs_guide.
+   *
+   * @param[in] initial var vector
+   * @param[in] param var vector
+   * @param[in] t time
+   * @param[in] y state of the base ODE system
+   * @param[in] yB state of the adjoint ODE system
+   * @param[out] yBdot evaluation of adjoint ODE RHS
+   */
+  inline void rhs_adj_sens(double t, N_Vector y, N_Vector yB,
+                           N_Vector yBdot) const {
+    Eigen::Map<const Eigen::VectorXd> y_vec(NV_DATA_S(y), N_);
+    Eigen::Map<Eigen::VectorXd> mu(NV_DATA_S(yB), N_);
+    Eigen::Map<Eigen::VectorXd> mu_dot(NV_DATA_S(yBdot), N_);
+    mu_dot.setZero();
+
+    const nested_rev_autodiff nested;
+
+    Eigen::Matrix<var, Eigen::Dynamic, 1> y_vars(y_vec);
+
+    Eigen::Matrix<var, Eigen::Dynamic, 1> f_y_t_vars
+        = rhs(t, y_vars, value_of_args_tuple_);
+
+    check_size_match(solver_->function_name_, "dy_dt", f_y_t_vars.size(),
+                     "states", N_);
+
+    f_y_t_vars.adj() = -mu;
+
+    grad();
+
+    mu_dot = y_vars.adj();
+  }
+
+  /*
+   * Calculate the RHS for the quadrature part of the adjoint ODE
+   * problem.
+   *
+   * This is the integrand of equation 2.22 in the cvs_guide.
+   *
+   * @param[in] initial var vector
+   * @param[in] param var vector
+   * @param[in] t time
+   * @param[in] y state of the base ODE system
+   * @param[in] yB state of the adjoint ODE system
+   * @param[out] qBdot evaluation of adjoint ODE quadrature RHS
+   */
+  inline void quad_rhs_adj(double t, N_Vector y, N_Vector yB, N_Vector qBdot) {
+    Eigen::Map<const Eigen::VectorXd> y_vec(NV_DATA_S(y), N_);
+    Eigen::Map<Eigen::VectorXd> mu(NV_DATA_S(yB), N_);
+    Eigen::Map<Eigen::VectorXd> mu_dot(NV_DATA_S(qBdot), num_args_vars_);
+    mu_dot.setZero();
+
+    const nested_rev_autodiff nested;
+
+    // The vars here do not live on the nested stack so must be zero'd
+    // separately
+    stan::math::for_each([](auto&& arg) { zero_adjoints(arg); },
+                         local_args_tuple_);
+
+    Eigen::Matrix<var, Eigen::Dynamic, 1> f_y_t_vars
+        = rhs(t, y_vec, local_args_tuple_);
+
+    check_size_match(solver_->function_name_, "dy_dt", f_y_t_vars.size(),
+                     "states", N_);
+
+    f_y_t_vars.adj() = -mu;
+
+    grad();
+
+    apply([&](auto&&... args) { accumulate_adjoints(mu_dot.data(), args...); },
+          local_args_tuple_);
+  }
+
+  /**
+   * Calculates the jacobian of the ODE RHS wrt to its states y at the
+   * given time-point t and state y.
+   */
+  inline void jacobian_states(double t, N_Vector y, SUNMatrix J) const {
+    Eigen::Map<Eigen::MatrixXd> Jfy(SM_DATA_D(J), N_, N_);
+    Eigen::Map<const Eigen::VectorXd> x(NV_DATA_S(y), N_);
+
+    nested_rev_autodiff nested;
+
+    const Eigen::Matrix<var, Eigen::Dynamic, 1> y_var(x);
```

Comment:
var types are never actually const

```suggestion
    Eigen::Matrix<var, Eigen::Dynamic, 1> y_var(x);
```